x-pack/plugins/cases/server/services/user_actions/index.ts

```diff
@@ -22,28 +42,282 @@ import {
+  cases: Array<{ id: string; owner: string; subCaseId?: string; connectorId: string }>;
+  user: User;
+}
+
+interface GetUserActionItemByDifference extends CommonUserActionArgs {
+  field: string;
+  originalValue: unknown;
+  newValue: unknown;
+}
+
+interface BulkCreateBulkUpdateCaseUserActions extends ClientArgs {
+  originalCases: Array<SavedObject<CaseAttributes | SubCaseAttributes>>;
+  updatedCases: Array<SavedObjectsUpdateResponse<CaseAttributes | SubCaseAttributes>>;
+  user: User;
+}
+
+interface BulkCreateAttachmentDeletionUserAction extends Omit<CommonUserActionArgs, 'owner'> {
+  attachments: Array<{ id: string; owner: string; attachment: CommentRequest }>;
+}
+
+type CreateUserActionClient<T extends keyof BuilderParameters> = CreateUserAction<T> &
+  CommonUserActionArgs;
+
 export class CaseUserActionService {
+  private static readonly userActionFieldsAllowed: Set<string> = new Set(Object.keys(ActionTypes));
+
+  private readonly builderFactory: BuilderFactory = new BuilderFactory();
+
   constructor(private readonly log: Logger) {}
 
+  private getUserActionItemByDifference({
+    field,
+    originalValue,
+    newValue,
+    caseId,
+    subCaseId,
+    owner,
+    user,
+  }: GetUserActionItemByDifference): BuilderReturnValue[] {
+    if (!CaseUserActionService.userActionFieldsAllowed.has(field)) {
+      return [];
+    }
+
+    if (field === ActionTypes.tags) {
+      const tagsUserActionBuilder = this.builderFactory.getBuilder(ActionTypes.tags);
+      const compareValues = isTwoArraysDifference(originalValue, newValue);
+      const userActions = [];
+
+      if (compareValues && compareValues.addedItems.length > 0) {
+        const tagAddUserAction = tagsUserActionBuilder?.build({
+          action: Actions.add,
+          caseId,
+          subCaseId,
+          user,
+          owner,
+          payload: { tags: compareValues.addedItems },
+        });
+
+        if (tagAddUserAction) {
+          userActions.push(tagAddUserAction);
+        }
+      }
+
+      if (compareValues && compareValues.deletedItems.length > 0) {
+        const tagsDeleteUserAction = tagsUserActionBuilder?.build({
+          action: Actions.delete,
+          caseId,
+          subCaseId,
+          user,
+          owner,
+          payload: { tags: compareValues.deletedItems },
+        });
+
+        if (tagsDeleteUserAction) {
+          userActions.push(tagsDeleteUserAction);
+        }
+      }
+
+      return userActions;
+    }
+
+    if (isUserActionType(field)) {
+      const userActionBuilder = this.builderFactory.getBuilder(ActionTypes[field]);
+      const fieldUserAction = userActionBuilder?.build({
+        caseId,
+        subCaseId,
+        owner,
+        user,
+        // @ts-ignore
+        payload: { [field]: newValue },
+      });
+
+      return fieldUserAction ? [fieldUserAction] : [];
+    }
+
+    return [];
+  }
+
+  public async bulkCreateCaseDeletionUserAction({
```

Comment:
nit: since we're in the user action service and generally whenever we call one of these methods it usually looks like `userActionService.<something>()` I think we can drop the `UserAction` part of each of these method names. I feel like it's redundant.